Sifchain/sa-eliza
# feat: github plugins

# Relates to:

Issue: Ensure dynamic parsing, validation, and execution for initializing repository, committing changes, creating pull requests, and managing memories from files using the GitHub API.

# Risks

**Risk Level**: Medium

- Impact on GitHub repository data (pull request, commits, etc.) due to incorrect owner, repository, or branch inputs.
- Dependencies on external Octokit and simple-git libraries for stability.

# Background

## What does this PR do?

This PR adds functionality to dynamically parse, validate, and execute GitHub operations such as initializing repositories, committing changes, creating pull requests, and managing data from files. It includes updates to ensure that inputs for repository owner, repository name, and branch are verified to minimize errors. Additionally, it introduces improved error handling for GitHub API calls.

## What kind of change is this?

- **Features**: Adds dynamic handling and validation for GitHub API operations. Introduces 4 new plugins: initialize repository, create commit, create pull request, and memories from files.
- **Improvements**: Enhances error handling and validation for repository management workflows.

# Documentation changes needed?

The PR changes include updates to the plugin document to explain how the new plugins can be used, including instructions on how to set the required GitHub API token.

# Testing

## Where should a reviewer start?

The reviewer should begin by checking the new GitHub API interaction methods, especially around initializing repositories, creating commits and creating pull requests. This includes verifying the correctness of the input validation functions.

## Detailed testing steps

1. Set up a test GitHub repository.
2. Set the GITHUB\_API\_TOKEN environment variable within the .env file.
3. Start the Eliza agent with a character file.
4. Trigger one of the plugin actions by using the relevant prompt to initialize a repository, commit changes, or create pull requests.
5. Verify that the correct data is passed, and that API calls return expected responses without errors.

No UI changes were made, and automated tests should be sufficient for verification.
